File: stacks_and_queues/stacks/array_inputs/c++/validate_stack/validate_stack.cpp
// validate_stack.cpp
#include "validate_stack.hpp"
#include <vector>
#include <stack>

bool validateStack(const std::vector<int> &pushed, const std::vector<int> &popped)
{
    if (pushed.size() == 1)
    {
        return pushed[0] == popped[0];
    }

    std::stack<int, std::vector<int>> stack;
    int poppedIndex = 0;
    for (auto &val : pushed)
    {
        // Push to stack
        stack.push(val);
        // As long as stack is non-empty and its top element is the same as the current popped element
        while (!stack.empty() && stack.top() == popped[poppedIndex])
        {
            // Pop the element from the stack
            stack.pop();
            // Increment the popped index to consider the next popped element
            ++poppedIndex;
        }
    }

    // If stack is empty, then the sequences of operations are valid, if not, then these sequences are not valid
    return stack.empty();
}

File: stacks_and_queues/stacks/array_inputs/python3/validate_stack.py
from typing import List

def validate_stack(pushed: List[int], popped: List[int]) -> bool:
    """
    Determine if `pushed` and `popped` can be obtained from a series of stack push and pop operations.

    Parameters
    ----------
    pushed : List[int]
        A series of pushed elements in the specified order
    popped : List[int]
        A series of popped elements in the specified order

    Returns
    -------
    bool
        `True` if the sequences of operations are valid, `False` otherwise
    """
    if len(pushed) == 1:
        return pushed[0] == popped[0]

    stack: List[int] = []
    pop_index = 0
    # For all elements in pushed
    for val in pushed:
        # Push to empty stack
        stack.append(val)
        # As long as stack is not empty and the top element is the same as current popped element
        while stack and stack[-1] == popped[pop_index]:
            # Pop the top of the stack
            stack.pop()
            # Increment the index to consider the next popped element
            pop_index += 1

    # If stack is empty, the sequence is valid, if not, it is not valid
    return not stack

def main() -> int:

    test_cases = [
        ([1, 2, 3, 4, 5], [4, 5, 3, 2, 1]),
        ([1, 2, 3, 4, 5], [4, 3, 5, 1, 2]),
        ([2], [3])
    ]
    for pushed, popped in test_cases:
        valid = validate_stack(pushed, popped)
        print(f"Given pushed = {pushed} and popped = {popped}, these sequences are {'valid' if valid else 'not valid'}")

    return 0

if __name__ == "__main__":

    main()
